dvc/repo/gc.py

```diff
@@ -82,23 +108,34 @@ def gc(  # noqa: PLR0913
                 jobs=jobs,
                 revs=[rev] if rev else None,
                 num=num or 1,
-            ).values():
-                used_obj_ids.update(obj_ids)
+            ).items():
+                odb_to_obj_ids.append((odb, obj_ids))
+
+    if not odb_to_obj_ids:
+        odb_to_obj_ids = [(None, set())]
+
+    if not_in_remote:
+        used_obj_ids = _used_obj_ids_not_in_remote(self, remote, jobs, odb_to_obj_ids)
+    else:
+        used_obj_ids = set()
+        for _, obj_ids in odb_to_obj_ids:
+            used_obj_ids.update(obj_ids)
 
     for scheme, odb in self.cache.by_scheme():
         if not odb:
             continue
-
         removed = ogc(odb, used_obj_ids, jobs=jobs)
         if not removed:
             logger.info("No unused '%s' cache to remove.", scheme)
 
     if not cloud:
         return
 
-    odb = self.cloud.get_remote_odb(remote, "gc -c")
-    removed = ogc(odb, used_obj_ids, jobs=jobs)
-    if removed:
-        get_index(odb).clear()
-    else:
-        logger.info("No unused cache to remove from remote.")
+    for remote_odb, obj_ids in odb_to_obj_ids:
```

Comment:
@dberenbaum I think it makes sense how this P.R. works for `remote:` + `--not-in-remote` but not sure about `--cloud`.

From the user's perspective, I don't really know what would be the expected/"right" behavior from mixing `remote: ` in `.dvc` files and `dvc gc -c`.

Reply:
tldr lgtm. As discussed, I think `remote:` doesn't make sense as a backup (it would require editing `remote:` frequently) but instead for cases where project data is split between different remotes (one dataset goes to a remote with wide access and another goes to a remote with very strict access). In the latter use case, I think it's expected that `dvc gc -c` would aggressively delete all files from a non-default remote except those specified in `remote:`. Of course, this could be dangerous and we should warn accordingly.